### PR TITLE
Clean up verification case 1b

### LINKED ISSUE #87: Issues with verification case 1.b
## Bug Description
1. MOOSE input file
    1. The [Line sampler](https://github.com/idaholab/TMAP8/blob/ffae64ced7169e725aed940d79ae99a7d8217a9c/test/tests/ver-1b/ver-1b.i#L57) vector PostProcessor should plot the full distance that we use in the verification with the appropriate resolution. Ideally the lineplot should only output on a specific timestep, since we are currently generating a very large number of line plot files.
    2. The PostProcessor [flux_point2](https://github.com/idaholab/TMAP8/blob/ffae64ced7169e725aed940d79ae99a7d8217a9c/test/tests/ver-1b/ver-1b.i#L73) should be at 0.5 m not 0.2 m, to be consistent with the documentation and Python script.
2. Python script
    1. Running the Python script against the gold csv file generates a plot where the analytical and numerical solutions don't match. The gold folder line plot matches the analytical equation at time 50 s, but the documentation and [Python script](https://github.com/idaholab/TMAP8/blob/c711bc80a10bddee5319d0d560a10a63dd5346e3/test/tests/ver-1b/comparison_ver-1b.py#L51) use 25 s. The Python script should be tested against the output from the actual input file to ensure transparency in our verification results.
### Comparsion against gold with time 25 seconds in analytical solution (As claimed in documentation)
ver-1b_comparison_dist_25s
### Comparsion against gold with time 50 seconds in analytical solution (What actually matches the gold csv plot)
ver-1b_comparison_dist_50s

## Steps to Reproduce
Run the MOOSE simulation and Python script as is from devel.

## Impact
Fixing this will ensure that the verification test is run accurately, and is run on an output files that the user can verify the inputs of. We shouldn't be running verifications on gold files as this creates a black box situation.

### PR DESCRIPTION
**Pull request by ChaitanyaBhave-e**
#closes #87

Fixes issues with verification case 1b - 

1. Input file now matches documentation and produces right outputs.
2. Only outputs vector PostProcessor at t=25 s, so that we don't clutter the folder with output csv files.
3. Python script fixed to use updated matplotlib syntax and now uses simulation result files instead of gold files for verification testing.

### CODE REVIEW

**Comment by ChaitanyaBhave-e** [comment]
@simopier @cticenhour

File: test/tests/ver-1b/comparison_ver-1b.py [hunk 1]
```diff
@@ -11,7 +11,7 @@
 gs = gridspec.GridSpec(1,1)
 ax = fig.add_subplot(gs[0])
 
-tmap_sol = pd.read_csv("./gold/u_vs_t.csv")
+tmap_sol = pd.read_csv("ver-1b_csv.csv")
```
Comment (simopier) [inline test/tests/ver-1b/comparison_ver-1b.py:14]: Why not use the gold file? Unless I am missing something, this is conflicting with the changes that Casey made in ver-1a to use the gold files. Using gold files helps to automatically create the documentation without running all the tests.

File: test/tests/ver-1b/comparison_ver-1b.py [hunk 2]
```diff
@@ -43,7 +43,7 @@
 gs = gridspec.GridSpec(1,1)
 ax = fig.add_subplot(gs[0])
 
-tmap_sol = pd.read_csv("./gold/u_vs_x.csv")
+tmap_sol = pd.read_csv("ver-1b_vector_postproc_line_0250.csv")
```
Comment (simopier) [inline test/tests/ver-1b/comparison_ver-1b.py:46]: Same here about gold file.

File: test/tests/ver-1b/comparison_ver-1b.py [hunk 3]
```diff
@@ -73,7 +73,7 @@
 gs = gridspec.GridSpec(1,1)
 ax = fig.add_subplot(gs[0])
 
-tmap_sol = pd.read_csv("./gold/u_vs_t.csv")
+tmap_sol = pd.read_csv("ver-1b_csv.csv")
```
Comment (simopier) [inline test/tests/ver-1b/comparison_ver-1b.py:76]: Same here for the gold file.

**Review by simopier — CHANGES_REQUESTED** [review]
This resolves the inconsistencies between the input file and the documentation. However, more work will need to happen in a follow up PR to clean up the documentation further, provide additional information, and automatically generate the documentation from the gold files, as was done in ver-1a. 

These comments are mostly related to the use of the test files rather than gold files in plotting comparisons.

File: test/tests/ver-1b/comparison_ver-1b.py (hunk 1 above)
Comment (ChaitanyaBhave-e) [reply test/tests/ver-1b/comparison_ver-1b.py:14]: I wanted to use the actual run solutions because I noticed that the gold file isn't always the same as what we claim on the documentation, and having the plots get generated from the run of an input file creates more transparency. I had mentioned this in the verification standards documentation too, I think. But your point makes sense, so I'm fine with reverting this change. In that case, we should make extra sure any PR that changes code/input files has to update the gold files and ensure they match what's in the documentation.

Comment (simopier) [reply test/tests/ver-1b/comparison_ver-1b.py:14]: Yes, absolutely. I think you should update the gold files in this PR, too.

Comment (ChaitanyaBhave-e) [reply test/tests/ver-1b/comparison_ver-1b.py:14]: I already updated the gold files anyway in this case because the old ones were wrong. But yes, I'll change the Python script to use gold again.

Comment (RemDelaporteMathurin) [reply test/tests/ver-1b/comparison_ver-1b.py:14]: I think we discussed that a while ago but ideally you wouldn't have gold files but a documentation that is generated on-the-fly

Comment (simopier) [reply test/tests/ver-1b/comparison_ver-1b.py:14]: @RemDelaporteMathurin, I'm not sure I understand your point about not needing gold files. 
Gold files in TMAP8 are used to store the tests/validation/examples results. They are updated to represent the latest version of TMAP8 predictions (otherwise tests fail). So having them is an integral part of quality control. 

As what we did with ver-1a, we will use these gold file (which contain the updated results) to generate the documentation. This would ensure that figures in the documentation are automatically generated by the latest TMAP8 version, which I think is what you were referring to.

Comment (RemDelaporteMathurin) [reply test/tests/ver-1b/comparison_ver-1b.py:14]: My point is that there is no reason to hard code these files since they can be generated by TMAP8 directly. As @ChaitanyaBhave-e pointed out:

> we should make extra sure any PR that changes code/input files has to update the gold files and ensure they match what's in the documentation

Which means that as soon as someone forgets this, then the documentation is out-of-date again...

Instead, the figure should be generated when compiling the documentation, not hard-coded.

Comment (RemDelaporteMathurin) [reply test/tests/ver-1b/comparison_ver-1b.py:14]: If you take a look at the [HTM documentation](https://h-transport-materials.readthedocs.io/en/latest/user/plotting.html#plotting-a-property) we have no hard coded data for the figures and they are generated on-the-fly when compiling the documentation. We don't have to "remember" to update the data for the figures at every PR

Comment (simopier) [reply test/tests/ver-1b/comparison_ver-1b.py:14]: If a PR changes code/input files, then tests will fail is the gold file is not updated as well, and the PR will not get merged until the gold files are updated (or if TMAP8 behavior was supposed to remain the same, then the change is corrected).

The gold files are therefore always reflecting the latest version of TMAP8, and tying the documentation to them ensures that the documentation is always reflecting the latest version of TMAP8. 

The documentation is here generated on the fly without having to run any of the tests.

Comment (simopier) [reply test/tests/ver-1b/comparison_ver-1b.py:14]: To be very clear: gold files are not hard-coded result files used only to create the documentation. 
Gold files are the reference files that define the gold standards for up-to-date TMAP8's predictions. 

When we perform tests in a PR (see at the bottom of the page), TMAP8 is compiled and run and its predictions are compared against the gold files. If a change affects TMAP8's behavior and the results do not match the gold file, then the tests will tell us that that specific test failed. To pass the test, the developer will have to update the old file, or fix the unexpected change in behavior. 
That's why gold files are the very best source of data to build an up-to-date documentation.

Running a fresh version of TMAP8 (without local changes) would therefore produce data that matches exactly the gold files (otherwise tests would have failed). Building the documentation on the gold files alleviate the need to run these tests (which can take a significant amount of time in the future as the number of tests/validation/example cases increases).

Comment (RemDelaporteMathurin) [reply test/tests/ver-1b/comparison_ver-1b.py:14]: Right so maybe my confusion comes from the fact that I thought TMAP8 transitioned to testing produced results against analytical data not previously (gold) produced data.

I understand what you mean.
I still believe that to be more rigorous the CI should compare the results against analytical solution produced on the fly (from the python functions) instead of comparing against previous versions of the code

Comment (simopier) [reply test/tests/ver-1b/comparison_ver-1b.py:14]: We do both.

The gold files are the reference files that define the gold standards for up-to-date TMAP8's predictions. They help us compare against previous versions of the code. This way, we are made aware of any changes that would affect TMAP8 predictions. 

Meanwhile, the documentation uses the python scripts to read the gold files AND calculate the analytical expression to generate a figure that contains both and keep an up to date documentation. So TMAP8 test results are compared both against the analytical solution (when it exists) AND the previous version of the code. 
This is all done properly for the ver-1a case. 

When no analytical solution is available, then we can use benchmarking results or experimental data instead (or sometimes there's nothing to compare against). In any case, new capabilities are always tested against TMAP8's previous version (via the gold files). This ensures robustness and stability.

File: test/tests/ver-1b/gold/u_vs_t.csv
Comment (ChaitanyaBhave-e) [inline test/tests/ver-1b/gold/u_vs_t.csv:1]: I realized this file did not get updated correctly. Fixing it in the updated push.

**Comment by ChaitanyaBhave-e** [comment]
@simopier ready for review

File: test/tests/ver-1b/gold/u_vs_t.csv
Comment (simopier) [reply test/tests/ver-1b/gold/u_vs_t.csv:1]: @ChaitanyaBhave-e 
Following up on the discussion with @RemDelaporteMathurin above, please add tests (in the `tests` file) that test the gold files `u_vs_t.csv` and `u_vs_x.csv`. These files should be automatically generated by ver-1b.i. They are currently not being tested/tracked and that's not right (for the reasons discussed above).

**Review by simopier — CHANGES_REQUESTED** [review]
> @ChaitanyaBhave-e
Following up on the discussion with @RemDelaporteMathurin above, please add tests (in the tests file) that test the gold files u_vs_t.csv and u_vs_x.csv. These files should be automatically generated by ver-1b.i. They are currently not being tested/tracked and that's not right (for the reasons discussed above).

File: test/tests/ver-1b/comparison_ver-1b.py (hunk 1 above)
Comment (simopier) [inline test/tests/ver-1b/comparison_ver-1b.py:14]:
```suggestion
tmap_sol = pd.read_csv("./gold/ver-1b_csv.csv")
```
We settled on using the gold files.

File: test/tests/ver-1b/comparison_ver-1b.py (hunk 2 above)
Comment (simopier) [inline test/tests/ver-1b/comparison_ver-1b.py:46]:
```suggestion
tmap_sol = pd.read_csv("./gold/ver-1b_vector_postproc_line_0250.csv")
```
We settled on using the gold files.

File: test/tests/ver-1b/comparison_ver-1b.py (hunk 3 above)
Comment (simopier) [inline test/tests/ver-1b/comparison_ver-1b.py:76]:
```suggestion
tmap_sol = pd.read_csv("./gold/ver-1b_csv.csv")
```

File: test/tests/ver-1b/tests [hunk 4]
```diff
@@ -1,11 +1,44 @@
...
+      exodiff = ver-1b_test_out.e
+      cli_args = 'Executioner/dt=10 Mesh/nx=100 Outputs/exodus/file_base=ver-1b_test_out'
+      requirement = 'The system shall be able to model transient diffusion through a slab with a constant concentration boundary condition as the species source.'
+    []
+    [ver-1b_heavy]
+      type = Exodiff
+      input = ver-1b.i
+      exodiff = ver-1b_out.e
+      heavy = true
+      requirement = 'The system shall be able to model transient diffusion through a slab with a constant concentration boundary condition as the species source.'
```
Comment (simopier) [inline test/tests/ver-1b/tests:16]:
```suggestion
      requirement = 'The system shall be able to model transient diffusion through a slab with a constant concentration boundary condition as the species source with the fine mesh and time step required to match the analytical solution.'
```

File: test/tests/ver-1b/tests [hunk 5]
```diff
@@ -1,11 +1,44 @@
...
+    []
+    [ver-1b_heavy_csvdiff]
+      type = CSVDiff
+      input = ver-1b.i
+      should_execute = False  # this test relies on the output files from ver-1b_heavy, so it shouldn't be run twice
+      csvdiff = ver-1b_csv.csv
+      requirement = 'The system shall be able to model transient diffusion through a slab with a constant concentration boundary condition as the species source, with the fine mesh and timestep required to match the analytical solution to generate CSV data for use in comparisons with the analytic solution.'
```
Comment (simopier) [inline test/tests/ver-1b/tests:23]:
```suggestion
      requirement = 'The system shall be able to model transient diffusion through a slab with a constant concentration boundary condition as the species source, with the fine mesh and time step required to match the analytical solution to generate CSV data for use in comparisons with the analytic solution over time.'
```

File: test/tests/ver-1b/tests [hunk 6]
```diff
@@ -1,11 +1,44 @@
...
+      heavy = true
+      prereq = ver-1b_heavy
+    []
+    [ver-1b_heavy_lineplot]
+      type = CSVDiff
+      input = ver-1b.i
+      should_execute = False  # this test relies on the output files from ver-1b_heavy, so it shouldn't be run twice
+      csvdiff = ver-1b_vector_postproc_line_0250.csv
+      requirement = 'The system shall be able to model transient diffusion through a slab with a constant concentration boundary condition as the species source, with the fine mesh and timestep required to match the analytical solution to generate CSV data for use in comparisons with the analytic solution.'
```
Comment (simopier) [inline test/tests/ver-1b/tests:32]:
```suggestion
      requirement = 'The system shall be able to model transient diffusion through a slab with a constant concentration boundary condition as the species source, with the fine mesh and timestep required to match the analytical solution to generate CSV data for use in comparisons with the analytic solution for the profile concentration.'
```
The test requirements must be unique (that is the current SQA documentation failure you are getting), and being more specific help with that.

**Review by simopier — CHANGES_REQUESTED** [review]
Great work! 
I still have a few comments, and then this case will still need to be cleaned up to follow the guidelines we developed, but that can be a separate PR.

**Review by simopier — APPROVED** [review]
Looking great!
